app/src/main/java/com/example/voyagerx/repository/LaunchRepository.kt

```diff
@@ -1,57 +1,129 @@
 package com.example.voyagerx.repository
 
+import android.content.Context
+import android.content.SharedPreferences
 import android.util.Log
 import com.example.rocketreserver.LaunchListQuery
+import com.example.voyagerx.R
 import com.example.voyagerx.repository.database.VoyagerXDatabase
 import com.example.voyagerx.repository.model.Launch
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.*
+import kotlinx.coroutines.Dispatchers.IO
 import java.util.*
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class LaunchRepository @Inject constructor(
     private val database: VoyagerXDatabase,
-    private val apiService: SpaceXApiService
+    private val apiService: SpaceXApiService,
+    @ApplicationContext private val context: Context
 ){
 
     suspend fun getLaunches() : List<Launch?>? {
-        Log.d("Repository", "attempting to fetch launch data")
+
+        val cachedLaunches = getLaunchesFromCache()
+        if(cachedLaunches != null){
+            Log.d("Repository", "get launch from cache")
+            return cachedLaunches
```

Comment:
This is interesting, I hadn't considered caching launches this way. Is there an advantage to using SharedPreferences as a cache vs checking for already inserted launches from the room database and returning those? Overall looks/works well! Just deciding on how or if we should update the cache I guess?

Reply:
I had initially thought about using the values from the database, but I was having difficulty implementing it and its still a fairly expensive operation to query from the db. I think this way is faster (?).

Reply:
Good point, I guess in theory we could benchmark the difference if this were a real production app but hey it's a cache. This should make the landing page much quicker, thank you!